In [30]:
import numpy as np
import pandas as pd

In [31]:
db = pd.read_csv("datasets/airbnb_nyc.csv")
db.head()

id                        listing_url  \
0  2595  https://www.airbnb.com/rooms/2595   
1  5121  https://www.airbnb.com/rooms/5121   
2  6848  https://www.airbnb.com/rooms/6848   
3  6990  https://www.airbnb.com/rooms/6990   
4  6872  https://www.airbnb.com/rooms/6872   

                                                name  \
0  Rental unit in New York · ★4.68 · Studio · 1 b...   
1  Rental unit in Brooklyn · ★4.52 · 1 bedroom · ...   
2  Rental unit in Brooklyn · ★4.58 · 2 bedrooms ·...   
3  Rental unit in New York · ★4.88 · 1 bedroom · ...   
4  Condo in New York · 1 bedroom · 1 bed · 1 shar...   

                                         picture_url  latitude  longitude  \
0  https://a0.muscache.com/pictures/miso/Hosting-...  40.75356  -73.98559   
1  https://a0.muscache.com/pictures/2090980c-b68e...  40.68535  -73.95512   
2  https://a0.muscache.com/pictures/e4f031a7-f146...  40.70935  -73.95342   
3  https://a0.muscache.com/pictures/be6cd5b3-9295...  40.78778  -73.94759   
4  https://a0.muscache.com/pictures/miso/Hosting-...  40.80107  -73.94255   

                 property_type        room_type  accommodates bathrooms_text  \
0           Entire rental unit  Entire home/apt             1         1 bath   
1  Private room in rental unit     Private room             2            NaN   
2           Entire rental unit  Entire home/apt             3         1 bath   
3  Private room in rental unit     Private room             1  1 shared bath   
4        Private room in condo     Private room             1  1 shared bath   

   bedrooms  beds                                          amenities    price  \
0       NaN   1.0  ["Long term stays allowed", "Cooking basics", ...  $240.00   
1       1.0   1.0  ["Air conditioning", "Wifi", "Kitchen", "Heati...   $66.00   
2       2.0   1.0  ["Cooking basics", "Microwave", "Extra pillows...   $81.00   
3       1.0   1.0  ["Hair dryer", "Air conditioning", "Fire extin...   $70.00   
4       1.0   1.0  ["Fire extinguisher", "Long term stays allowed...   $65.00   

   minimum_nights  maximum_nights  availability_30  availability_365  \
0              30            1125                4               339   
1              30             730               30                88   
2              30             730                0               142   
3              30             365                0                93   
4              30             180               24               359   

  neighbourhood_group       neighbourhood  
0           Manhattan             Midtown  
1            Brooklyn  Bedford-Stuyvesant  
2            Brooklyn        Williamsburg  
3           Manhattan         East Harlem  
4           Manhattan         East Harlem

In [32]:
db.describe()

id      latitude     longitude  accommodates      bedrooms  \
count  3.962700e+04  39627.000000  39627.000000  39627.000000  36218.000000   
mean   3.083450e+17     40.729360    -73.946072      2.874833      1.435336   
std    4.050888e+17      0.056425      0.054788      2.079134      0.866508   
min    2.595000e+03     40.500314    -74.251907      1.000000      1.000000   
25%    2.023932e+07     40.688980    -73.982785      2.000000      1.000000   
50%    4.650264e+07     40.726520    -73.954090      2.000000      1.000000   
75%    7.564324e+17     40.762667    -73.926560      4.000000      2.000000   
max    1.037983e+18     40.911380    -73.713650     16.000000     26.000000   

               beds  minimum_nights  maximum_nights  availability_30  \
count  38972.000000    39627.000000    3.962700e+04     39627.000000   
mean       1.642769       30.051127    5.581488e+04        10.512984   
std        1.169641       29.290027    1.078876e+07        12.643032   
min        1.000000        1.000000    1.000000e+00         0.000000   
25%        1.000000       30.000000    1.800000e+02         0.000000   
50%        1.000000       30.000000    3.650000e+02         2.000000   
75%        2.000000       30.000000    1.125000e+03        27.000000   
max       42.000000     1250.000000    2.147484e+09        30.000000   

       availability_365  
count      39627.000000  
mean         153.721276  
std          147.391076  
min            0.000000  
25%            0.000000  
50%          122.000000  
75%          319.000000  
max          365.000000

1. Handling missing values:
The first step is to handle missing values. We can do this by identifying how many missing values are present in each column. Then we will be handling them one by one.

In [33]:
null = db.isnull()
print(null.sum())

id                        0
listing_url               0
name                      0
picture_url               0
latitude                  0
longitude                 0
property_type             0
room_type                 0
accommodates              0
bathrooms_text           37
bedrooms               3409
beds                    655
amenities                 0
price                     0
minimum_nights            0
maximum_nights            0
availability_30           0
availability_365          0
neighbourhood_group       0
neighbourhood             0
dtype: int64


We have 37, 3409 and 655 missing values in bathroom_text, bathrooms and beds. We would need to handle these values by removing the rows which have these following values. Another way to handle missing values can be by imputing values eg finding the average of the values and assigning it to the missing values, but since that might decrease the quality of the data in this case we will be removing rows completely which contain null values.

In [34]:
db.dropna(inplace=True)
null = db.isnull()
print(null.sum())

id                     0
listing_url            0
name                   0
picture_url            0
latitude               0
longitude              0
property_type          0
room_type              0
accommodates           0
bathrooms_text         0
bedrooms               0
beds                   0
amenities              0
price                  0
minimum_nights         0
maximum_nights         0
availability_30        0
availability_365       0
neighbourhood_group    0
neighbourhood          0
dtype: int64


2. Handling outliers: 
The next step is to handle outliers. We will be using Z score to identify the outliers.

In [35]:
from scipy.stats import zscore
numeric_columns = ['latitude', 'longitude','accommodates', 'bedrooms', 'beds','minimum_nights', 'maximum_nights', 'availability_30', 'availability_365']
z_scores = zscore(db[numeric_columns])
outliers = db[np.abs(z_scores) > 3].sum()
print(outliers)

id                              0.0
listing_url                       0
name                              0
picture_url                       0
latitude                 2482.54463
longitude             -68279.979369
property_type                     0
room_type                         0
accommodates                 8192.0
bathrooms_text                    0
bedrooms                     2140.0
beds                         4523.0
amenities                         0
price                             0
minimum_nights              71503.0
maximum_nights           40000000.0
availability_30                 0.0
availability_365                0.0
neighbourhood_group               0
neighbourhood                     0
dtype: object


After identifying the outliers we can either remove the rows or can cap the values of the outliers. In this case we will be capping the values of the outliers, so that we can include the rest of the row.

In [36]:
threshold = 3
for col in numeric_columns:
    mean = db[col].mean()
    std = db[col].std()
    lowerbound = mean - threshold * std
    upperbound = mean + threshold * std
    db[col] = np.clip(db[col], lowerbound, upperbound)

z_scores_after_capping = zscore(db[numeric_columns])
outliers_after_capping = db[np.abs(z_scores_after_capping) > 3].sum()
print(outliers_after_capping)

id                               0.0
listing_url                        0
name                               0
picture_url                        0
latitude                 2524.143922
longitude              -77449.026571
property_type                      0
room_type                          0
accommodates             6839.758504
bathrooms_text                     0
bedrooms                 5195.200119
beds                     5684.790152
amenities                          0
price                              0
minimum_nights           79307.70507
maximum_nights         902363.036173
availability_30                  0.0
availability_365                 0.0
neighbourhood_group                0
neighbourhood                      0
dtype: object


We will be deleting the remaining rows with outliers

In [37]:
outlier_rows = set()
for col in numeric_columns:
     z_scores = zscore(db[col])
     outlier_rows.update(db[np.abs(z_scores)>threshold].index)

db_cleaned = db.drop(index = outlier_rows)

In [38]:
z_score_after_deleting = zscore(db_cleaned[numeric_columns])
outliers = db[np.abs(z_score_after_deleting) > threshold].sum()
print(outliers)

id                              0.0
listing_url                       0
name                              0
picture_url                       0
latitude                 163.258455
longitude             -42089.323032
property_type                     0
room_type                         0
accommodates                 3056.0
bathrooms_text                    0
bedrooms                     5259.0
beds                         4212.0
amenities                         0
price                             0
minimum_nights              10657.0
maximum_nights              52997.0
availability_30                 0.0
availability_365                0.0
neighbourhood_group               0
neighbourhood                     0
dtype: object


3. We will be checking for duplicate values. In this example we will be using .duplicated function to check for duplicate values. If there are duplicate values we will be dropping the rows with duplicate values. In this dataset we do not have duplicate values.

In [39]:
duplicates = db_cleaned["id"].duplicated()
print(duplicates.sum())

0


4. We will be looking for inconsistent values. To do this we will use .head to get a first five rows of the data. After checking the data we can confirm that there exist no inconsistency in the data like spaces, wrong casing etc.

In [40]:
db_cleaned.head()

id                        listing_url  \
2  6848  https://www.airbnb.com/rooms/6848   
3  6990  https://www.airbnb.com/rooms/6990   
4  6872  https://www.airbnb.com/rooms/6872   
5  7097  https://www.airbnb.com/rooms/7097   
6  7064  https://www.airbnb.com/rooms/7064   

                                                name  \
2  Rental unit in Brooklyn · ★4.58 · 2 bedrooms ·...   
3  Rental unit in New York · ★4.88 · 1 bedroom · ...   
4  Condo in New York · 1 bedroom · 1 bed · 1 shar...   
5  Home in Brooklyn · ★4.88 · 1 bedroom · 2 beds ...   
6  Loft in Brooklyn · ★4.91 · 1 bedroom · 1 bed ·...   

                                         picture_url  latitude  longitude  \
2  https://a0.muscache.com/pictures/e4f031a7-f146...  40.70935  -73.95342   
3  https://a0.muscache.com/pictures/be6cd5b3-9295...  40.78778  -73.94759   
4  https://a0.muscache.com/pictures/miso/Hosting-...  40.80107  -73.94255   
5  https://a0.muscache.com/pictures/miso/Hosting-...  40.69194  -73.97389   
6  https://a0.muscache.com/pictures/13708959/7e74...  40.71248  -73.95881   

                 property_type        room_type  accommodates  bathrooms_text  \
2           Entire rental unit  Entire home/apt           3.0          1 bath   
3  Private room in rental unit     Private room           1.0   1 shared bath   
4        Private room in condo     Private room           1.0   1 shared bath   
5         Private room in home     Private room           2.0  1 private bath   
6         Private room in loft     Private room           2.0   1 shared bath   

   bedrooms  beds                                          amenities    price  \
2       2.0   1.0  ["Cooking basics", "Microwave", "Extra pillows...   $81.00   
3       1.0   1.0  ["Hair dryer", "Air conditioning", "Fire extin...   $70.00   
4       1.0   1.0  ["Fire extinguisher", "Long term stays allowed...   $65.00   
5       1.0   2.0  ["Outdoor dining area", "Long term stays allow...  $205.00   
6       1.0   1.0  ["Air conditioning", "Iron", "Heating", "Essen...   $95.00   

   minimum_nights  maximum_nights  availability_30  availability_365  \
2            30.0           730.0                0               142   
3            30.0           365.0                0                93   
4            30.0           180.0               24               359   
5             2.0          1125.0                5               276   
6            30.0            45.0                0                26   

  neighbourhood_group neighbourhood  
2            Brooklyn  Williamsburg  
3           Manhattan   East Harlem  
4           Manhattan   East Harlem  
5            Brooklyn   Fort Greene  
6            Brooklyn  Williamsburg

4. We will be looking for wrong data types. To do this we will use .dtypes to get a list of columns and their datatypes. We would need to convert the price column data type from object to integer, since it might be required as a numeric column later or during the exploratory analysis.

In [41]:
print(db.dtypes)

id                       int64
listing_url             object
name                    object
picture_url             object
latitude               float64
longitude              float64
property_type           object
room_type               object
accommodates           float64
bathrooms_text          object
bedrooms               float64
beds                   float64
amenities               object
price                   object
minimum_nights         float64
maximum_nights         float64
availability_30          int64
availability_365         int64
neighbourhood_group     object
neighbourhood           object
dtype: object


FIrst we will be striping the $ sign off, then we will be converting the values to numeric. We will be making sure that the errors (which are the null values) get handled. We will then drop the rows which have null values. Finally we will be converting the price to int data type.

In [28]:
db_cleaned['price'] = db_cleaned['price'].str.strip('$')
db_cleaned['price'] = pd.to_numeric(db_cleaned['price'], errors= 'coerce')
db_cleaned = db_cleaned.dropna(subset=['price'])
db_cleaned['price'] = db_cleaned['price'].astype('int')

In [42]:
db_cleaned.describe()

id      latitude     longitude  accommodates      bedrooms  \
count  3.193100e+04  31931.000000  31931.000000  31931.000000  31931.000000   
mean   3.059156e+17     40.731041    -73.948316      2.669694      1.293163   
std    4.049394e+17      0.056112      0.045134      1.495324      0.563052   
min    6.848000e+03     40.559700    -74.103326      1.000000      1.000000   
25%    1.913378e+07     40.689870    -73.982120      2.000000      1.000000   
50%    4.557994e+07     40.727380    -73.953540      2.000000      1.000000   
75%    7.526935e+17     40.763444    -73.927310      4.000000      1.000000   
max    1.037983e+18     40.899759    -73.785890      8.000000      3.000000   

               beds  minimum_nights  maximum_nights  availability_30  \
count  31931.000000    31931.000000    31931.000000     31931.000000   
mean       1.505715       27.370737      618.126930        10.399424   
std        0.789211        8.924061      480.870684        12.616301   
min        1.000000        1.000000        1.000000         0.000000   
25%        1.000000       30.000000      120.000000         0.000000   
50%        1.000000       30.000000      365.000000         1.000000   
75%        2.000000       30.000000     1125.000000        26.000000   
max        4.000000       70.000000    10000.000000        30.000000   

       availability_365  
count      31931.000000  
mean         148.522408  
std          146.971685  
min            0.000000  
25%            0.000000  
50%           96.000000  
75%          308.000000  
max          365.000000

These are the steps for cleaning and pre-processing data for this dataset. In this project we used some basic data cleaning techniques to clean the data.